lucene/core/src/java/org/apache/lucene/store/EncryptingIndexInput.java

```diff
@@ -0,0 +1,299 @@
+  }
+
+  private EncryptingIndexInput(String resourceDescription, long sliceOffset, long sliceLength, boolean isClone,
+                               IndexInput indexInput, Key key, byte[] initialIv) throws IOException {
+    super(resourceDescription);
+    assert sliceOffset >= 0 && sliceLength >= 0;
+    this.sliceOffset = sliceOffset;
+    this.sliceEnd = sliceOffset + sliceLength;
+    this.isClone = isClone;
+    this.indexInput = indexInput;
+    this.key = key;
+    this.initialIv = initialIv;
+
+    cipher = createAesCtrCipher();
+    assert cipher.getBlockSize() == AES_BLOCK_SIZE : "Invalid AES block size: " + cipher.getBlockSize();
+    iv = initialIv.clone();
+    ivParameterSpec = new ReusableIvParameterSpec(iv);
+    try {
+      cipher.init(Cipher.DECRYPT_MODE, this.key, ivParameterSpec);
+    } catch (InvalidKeyException | InvalidAlgorithmParameterException e) {
+      throw new IOException(e);
+    }
+
+    inBuffer = ByteBuffer.allocate(getBufferSize());
+    outBuffer = ByteBuffer.allocate(getBufferSize() + AES_BLOCK_SIZE);
+    outBuffer.limit(0);
+    assert inBuffer.hasArray() && outBuffer.hasArray();
+    assert inBuffer.arrayOffset() == 0;
+    inArray = inBuffer.array();
+    oneByteBuf = new byte[1];
+  }
+
+  /**
+   * Reads the initial IV at the beginning of the index input.
+   */
+  private static byte[] readInitialIv(IndexInput indexInput) throws IOException {
+    indexInput.seek(HEADER_LENGTH);
+    byte[] initialIv = new byte[IV_LENGTH];
+    indexInput.readBytes(initialIv, 0, initialIv.length, false);
+    return initialIv;
+  }
+
+  /**
+   * Gets the buffer size. It must be a multiple of {@link EncryptingUtil#AES_BLOCK_SIZE}.
+   */
+  protected int getBufferSize() {
+    return BUFFER_SIZE;
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (!closed) {
+      closed = true;
+      if (!isClone) {
+        indexInput.close();
+      }
+    }
+  }
+
+  @Override
+  public long getFilePointer() {
+    return getPosition() - sliceOffset;
+  }
+
+  /**
+   * Gets the current internal position in the delegate {@link IndexInput}. It includes the initial IV length.
+   */
+  private long getPosition() {
+    return indexInput.getFilePointer() - outBuffer.remaining();
+  }
+
+  @Override
+  public void seek(long position) throws IOException {
+    if (position < 0) {
+      throw new IllegalArgumentException("Invalid position=" + position);
+    }
+    if (position > length()) {
+      throw new EOFException("Seek beyond EOF (position=" + position + ", length=" + length() + ") in " + this);
+    }
+    long targetPosition = position + sliceOffset;
+    long delegatePosition = indexInput.getFilePointer();
+    long currentPosition = delegatePosition - outBuffer.remaining();
+    if (targetPosition >= currentPosition && targetPosition <= delegatePosition) {
+      outBuffer.position(outBuffer.position() + (int) (targetPosition - currentPosition));
```

Comment:
Math.toIntExact()?

Reply:
The target position is within the buffered output, which has a small integer size. I'll add a comment here.